src/tsf/Implementation.cpp

```diff
@@ -0,0 +1,660 @@
+
+    std::wstring finalizedString;
+    std::wstring activeComposition;
+    til::small_vector<Render::CompositionRange, 2> activeCompositionRanges;
+    bool firstRange = true;
+
+    const GUID* guids[] = { &GUID_PROP_COMPOSING, &GUID_PROP_ATTRIBUTE };
+    wil::com_ptr<ITfReadOnlyProperty> props;
+    THROW_IF_FAILED(_context->TrackProperties(&guids[0], ARRAYSIZE(guids), nullptr, 0, props.addressof()));
+
+    wil::com_ptr<IEnumTfRanges> enumRanges;
+    THROW_IF_FAILED(props->EnumRanges(ec, enumRanges.addressof(), fullRange.get()));
+
+    // IEnumTfRanges::Next returns S_FALSE when it has reached the end of the list.
+    // This includes any call where the number of returned items is less than what was requested.
+    for (HRESULT nextResult = S_OK; nextResult == S_OK;)
+    {
+        ITfRange* ranges[8];
+        ULONG rangesCount;
+        nextResult = enumRanges->Next(ARRAYSIZE(ranges), &ranges[0], &rangesCount);
+
+        const auto cleanup = wil::scope_exit([&] {
+            for (ULONG i = 0; i < rangesCount; ++i)
+            {
+                ranges[i]->Release();
+            }
+        });
+
+        for (ULONG i = 0; i < rangesCount; ++i)
+        {
+            const auto range = ranges[i];
+
+            bool composing = false;
+            TfGuidAtom atom = TF_INVALID_GUIDATOM;
+            {
+                wil::unique_variant var;
+                THROW_IF_FAILED(props->GetValue(ec, range, var.addressof()));
+
+                wil::com_ptr<IEnumTfPropertyValue> propVal;
+                wil::com_query_to(var.punkVal, propVal.addressof());
+
+                unique_tf_propertyval propVals[2];
+                THROW_IF_FAILED(propVal->Next(2, propVals[0].addressof(), nullptr));
+
+                for (const auto& val : propVals)
+                {
+                    if (IsEqualGUID(val.guidId, GUID_PROP_COMPOSING))
+                    {
+                        composing = V_VT(&val.varValue) == VT_I4 && V_I4(&val.varValue) != 0;
+                    }
+                    else if (IsEqualGUID(val.guidId, GUID_PROP_ATTRIBUTE))
+                    {
+                        atom = V_VT(&val.varValue) == VT_I4 ? static_cast<TfGuidAtom>(V_I4(&val.varValue)) : TF_INVALID_GUIDATOM;
+                    }
+                }
+            }
+
+            size_t totalLen = 0;
+            for (;;)
+            {
+                static constexpr ULONG bufCap = 128;
+                WCHAR buf[bufCap];
+                ULONG len = bufCap;
+                THROW_IF_FAILED(range->GetText(ec, TF_TF_MOVESTART, buf, len, &len));
+
+                if (!composing && firstRange)
+                {
+                    finalizedString.append(buf, len);
+                }
+                else
+                {
+                    activeComposition.append(buf, len);
+                }
+
+                totalLen += len;
+
+                if (len < bufCap)

```

Comment:
what happens if we compose exactly 128 chars?

Reply:
(will we throw next time we `GetText`?)

Reply:
I'll test it and comment my results in the code.